pydomcfg/domzgr/zgr.py

```diff
@@ -0,0 +1,161 @@
+"""
+Base class to generate NEMO v4.0 vertical grids.
+"""
+
+from itertools import product
+
+import xarray as xr
+from xarray import DataArray, Dataset
+
+
+class Zgr:
+    """
+    Base class to generate NEMO vertical grids.
+    """
+
+    def __init__(self, ds_bathy: Dataset, jpk: int):
+        """
+        Initialize class.
+
+        Parameters
+        ----------
+        ds_bathy: Dataset
+             xarray dataset including grid coordinates and bathymetry
+        jpk: int
+             number of computational levels
+        """
+
+        self._bathy = ds_bathy
+        self._jpk = jpk
+
+    # -------------------------------------------------------------------------
+    def _init_ds(self):
+        """
+        Initialise the xarray dataset with empty
+        ``z3{T,W}`` and ``e3{T,W}``
+
+        Returns
+        -------
+        ds: Dataset
+            A copy of the dataset used to initialise the class with new
+            coordinates ``z3{T,W}`` and ``e3{T,W}`` empty dataarrays
+
+        """
+        ds = self._bathy.copy()
+
+        var = ["z3", "e3"]
+        grd = ["T", "W"]
+
+        # Initialise a dataset with empty z3 and e3 dataarrays
+        da = xr.full_like(ds["Bathymetry"], None).expand_dims(z=range(self._jpk))
+        for v, g in product(var, grd):
+            ds[v + g] = da.copy()
+            ds = ds.set_coords(v + g)
+        return ds
+
+    # -------------------------------------------------------------------------
+    def _sigma(self, kindx: DataArray, grd: str):
+        """
+        Provide the analytical function for sigma-coordinate,
+        a uniform non-dimensional vertical coordinate describing
+        the non-dimensional position of model levels.
+
+        Consider that -1 <= sigma <= 0, with
+
+                 sigma =  0 at the shallower boundary
+                 sigma = -1 at the deeper boundary
+
+        Parameters
+        ----------
+        kindx: DataArray
+            Model levels indexes. Note that
+            *) T-points are at integer values (between 1 and jpk)
+            *) W-points are at integer values - 1/2 (between 0.5 and jpk-0.5)
+        grd: str
+            If we are dealing with "T" or "W" model levels.
+
+        Returns
+        -------
+        ps: DataArray
+            Uniform non-dimensional sigma-coordinate (-1 <= sigma <= 0)
+        """
+
+        kindx = kindx + 1.0  # Fortran indexing
+
+        if grd == "W":
+            kindx -= 0.5
+
+        ps = -(kindx - 0.5) / float(self._jpk - 1.0)
+        return ps
```

Comment:
Looks like the two staggered grids are always coupled and called together. Does it make sense to get rid of `grd` and return a tuple? E.g., 
```python
        kindx = kindx + 1.0  # Fortran indexing
        
        return  (-(kindx - shift) / float(self._jpk - 1.0) for shift in (0.5, 1))
```

Reply:
Agreed!